srcs/utils.py

```diff
@@ -235,3 +216,176 @@ def write_processed_indices(filename, signatures, uncorr_org_idx):
+
+        # parse header with column indices
+        if line.startswith("@@"):
+            for index, column_name in enumerate(line[2:].split('\t')):
+                column_name_to_index[column_name] = index
+            index_rank, index_taxid, index_percentage, index_taxpath, index_taxpathsn = get_column_indices(column_name_to_index)
+            got_column_indices = True
+            reading_data = False
+            continue
+
+        # parse header with metadata
+        if line.startswith("@"):
+            # if last line contained sample data and new header starts, store profile for sample
+            if reading_data:
+                if 'SAMPLEID' in header and 'VERSION' in header and 'RANKS' in header:
+                    if len(profile) > 0:
+                        samples_list.append((header['SAMPLEID'], header, profile))
+                        profile = []
+                        predictions_dict = {}
+                else:
+                    logger.error("Header in file {} is incomplete. Check if the header of each sample contains at least SAMPLEID, VERSION, and RANKS.\n".format(file_path))
+                    raise RuntimeError
+                header = {}
+            reading_data = False
+            got_column_indices = False
+            key, value = line[1:].split(':', 1)
+            header[key.upper()] = value.strip()
+            continue
+
+        if not got_column_indices:
+            logger.error("Header line starting with @@ in file {} is missing or at wrong position.\n".format(file_path))
+            raise RuntimeError
+
+        reading_data = True
+        row_data = line.split('\t')
+
+        taxid = row_data[index_taxid]
+        # if there is already a prediction for taxon, only sum abundance
+        if taxid in predictions_dict:
+            prediction = predictions_dict[taxid]
+            prediction.percentage += float(row_data[index_percentage])
+        else:
+            if float(row_data[index_percentage]) == .0:
+                continue
+            prediction = Prediction()
+            predictions_dict[taxid] = prediction
+            prediction.taxid = row_data[index_taxid]
+            prediction.rank = row_data[index_rank]
+            prediction.percentage = float(row_data[index_percentage])
+            prediction.taxpath = row_data[index_taxpath]
+            if isinstance(index_taxpathsn, int):
+                prediction.taxpathsn = row_data[index_taxpathsn]
+            else:
+                prediction.taxpathsn = None
+            profile.append(prediction)
+
+    # store profile for last sample
+    if 'SAMPLEID' in header and 'VERSION' in header and 'RANKS' in header:
+        if reading_data and len(profile) > 0:
+            samples_list.append((header['SAMPLEID'], header, profile))
+    else:
+        logger.error("Header in file {} is incomplete. Check if the header of each sample contains at least SAMPLEID, VERSION, and RANKS.\n".format(file_path))
```

Comment:
<img width="498" alt="image" src="https://github.com/KoslickiLab/YACHT/assets/144844529/3787231d-f7c0-43f1-a2de-2731352fde55">
@chunyuma could you please fix? Same 3 times in this file

Reply:
Thanks @mfl15. I fixed it https://github.com/KoslickiLab/YACHT/commit/3a06a77de9d3675edf18cf5aa8cd26871b2194aa